Enable clang-tidy and refactor NULL to nullptr

jj-marr opened the pull request:
clang-tidy is a linter tool that automatically checks for common code quality issues. This PR adds a cmake option `DCLANG_TIDY` that when set with `DCLANG_TIDY=ON` will automatically lint Wesnoth's code for the issues specified in `src/.clang-tidy`

As a representative example, the `NULL` macro is an implementation-defined keyword that is typically the integer 0. This causes issues because it is not actually a pointer type, and so will trigger integer overloads of methods preferentially to pointer overloads. In C++11, the keyword `nullptr` was introduced which is actually a pointer type and doesn't have this problem. This PR removes the last remaining usages of `NULL` or `0` from the codebase and enables `clang-tidy` to check for usage of `NULL` automatically.

Git submodules not owned by `wesnoth` are excluded from linting.

To use clang-tidy with an in-tree build:

1. Install `clang` and tooling
2. `cmake . -DCLANG_TIDY=ON && make`

soliton- commented on src/CMakeLists.txt:
```diff
@@ -104,6 +104,8 @@ if(NOT ENABLE_SYSTEM_LUA)
 		set_property(SOURCE SOURCE ${lua_sources} APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-string-plus-int")
 	endif()
+
+	set_source_files_properties(${lua_sources} PROPERTIES SKIP_LINTING ON)
```
If this just skips `src/modules/lua` then it should skip `src/modules` instead. Any code under `src/modules` is not from us.

jj-marr replied:
Strictly speaking, this doesn't skip src/modules/lua, it sets the SKIP_LINTING attribute on all of the source files from lua that we're compiling.

I refactored the CMakeLists.txt to allow for enabling/disabling linting based on the directory.

Pentarctagon commented:
This should presumably be added as a build option for CI? Or maybe in #9355?

jj-marr commented:
I gave up on #9355 for now, because the dereferencing stuff is harder to resolve than autoreplacing `NULL` with `nullptr`.

Also, the way this is set up, `clang-tidy` runs incrementally whenever a source file is recompiled. If you want to lint the entire codebase, it's necessary to clean the build.

Pentarctagon commented:
You'd add it to the compile flags for the linux jobs in https://github.com/wesnoth/wesnoth/blob/master/.github/workflows/ci-main.yml. The additional compile option should also be added for scons as well.

CelticMinstrel commented:
> This causes issues because it is not actually a pointer type

I think it's typically defined as `(void*)0` actually… but that might not be guaranteed by standards.

CelticMinstrel commented on src/config_cache.cpp:
```diff
@@ -409,7 +409,7 @@ bool config_cache::delete_cache_files(const std::vector<std::string>& paths,
 
 config_cache_transaction::state config_cache_transaction::state_ = FREE;
-config_cache_transaction* config_cache_transaction::active_ = 0;
+config_cache_transaction* config_cache_transaction::active_ = nullptr;
```
What on earth is `= FREE`…?

jj-marr replied:
Very good question. I have no idea, I didn't write that line.

CelticMinstrel replied:
I assume it's defined as a macro or a constant somewhere in the source…

jj-marr replied:
It's an enum. I think it means that by default, there is no active cache transaction (i.e. nullptr) and the enum means it's free to take another transaction.

CelticMinstrel replied:
Ah, that makes sense.

CelticMinstrel commented:
It's nice that this also catches cases of using literal 0 as a pointer.

gfgtdf commented:
> I think it's typically defined as `(void*)0` actually… but that might not be guaranteed by standards.

No https://en.cppreference.com/w/cpp/types/NULL

CelticMinstrel commented:
Well, whatever the standard says, I'm pretty sure it's still defined as `(void*)0` on some implementations. Though for C++ it's usually defined as `nullptr` these days.

CelticMinstrel commented:
I don't have anything to say on the build system changes (someone else should probably review them) but all the source changes look good.

jj-marr commented:
@Pentarctagon The version of `scons` used in Wesnoth CI is 3.02, which doesn't support `compilation_db`. I'm not really sure how to update that. For the issue with getting this to work with USE_SYSTEM_LUA, I'm going to try to figure this out.

Pentarctagon commented:
What version is needed?

jj-marr commented:
4.0 or above.

Pentarctagon commented:
Which job is on 3.02? scons 4+ should be present.

jj-marr commented:
@Pentarctagon The Flatpak builder is on scons 3.02. Also, the scons gcc builder will require clang-tools to be installed in order to make the `run-clang-tidy` script and `clang-tidy` executable available.

soliton- commented:
For the CI it should be enough if one build does the clang-tidy checks unless it makes a difference to do it with optimizations and without or so.

Pentarctagon commented:
Looks like the scons version is from https://github.com/wesnoth/wesnoth/blob/master/packaging/flatpak/org.wesnoth.Wesnoth.json#L53.

jj-marr commented:
@Pentarctagon What should I do about the `scons` docker container not having `clang-tidy`? How would I modify that?

Pentarctagon commented:
Unless there's a specific need for it to run with the flatpak job, I'd say it's fine to just disable it there.

jj-marr commented:
Turns out `clang-tidy` wasn't installed in the docker environment so `clang-tidy` was never actually running in CI.

jj-marr commented:
Anyways, I finally got `clang-tidy` working on `scons` and `cmake` after several tries.

loonycyborg reviewed (approved):
Potentially can also integrate clang-tidy call into the scons build using AddPostAction or Command. What are clang-tidy's exact inputs?

jj-marr commented:
> What are clang-tidy's exact inputs?

In general, it only requires `compile_commands.json` to be generated. In Wesnoth specifically, some header files are autogenerated through mechanisms in the build system I don't understand. That's why `run-clang-tidy` is executed after the compilation is entirely done.

loonycyborg commented:
Changing your Alias to
 ```
Alias('cdb', cdb, "run-clang-tidy", ENV = os.environ)
```
seems to work ok. It would run clang-tidy directly but name of alias becomes a bit misleading then. I guess it's possible to add Depends() on header that needs to exist to avoid need of a separate call to entire build too. But I can look into this after this pr is merged.

Vultraz commented on .github/workflows/ci-main.yml:
```diff
@@ -104,7 +109,7 @@ jobs:
                   -DEXTRA_FLAGS_CONFIG="-pipe" -DENABLE_MYSQL=true \
                   -DENABLE_NLS=false -DENABLE_LTO="$LTO" -DFORCE_COLOR_OUTPUT=true -DLTO_JOBS=2 \
-                  -DENABLE_SYSTEM_LUA="$SYS_LUA" .
+                  -DENABLE_SYSTEM_LUA="$SYS_LUA -DCLANG_TIDY=true" .
```
Isn't there a bad quote here? Shouldn't it be `-DENABLE_SYSTEM_LUA="$SYS_LUA" -DCLANG_TIDY=true`

jj-marr commented:
Fine with squashing and merging this.

soliton- commented:
The addition of another `run-clang-tidy` call before building the compilation database just looks wrong.

Also cmake now seems to be confused about the compiler in use. It gives clang `-flto=2` which is fine for gcc but not clang.

jj-marr commented:
I added `-wno-unknown-warning-option` to the clang-tidy command line and removed the extraneous `run-clang-tidy` call. I've also refactored the PR into 5 commits to avoid one big squash-merge.

jj-marr commented on SConstruct:
```diff
@@ -189,6 +190,12 @@ if env['distcc']:
+if env['compile_db']:
+    env.Tool('compilation_db')
+    cdb = env.CompilationDatabase()
+    env['COMPILATIONDB_PATH_FILTER'] = "build/" + env["build"] + "/[!m][!o][!d]*"
```
https://docs.python.org/3/library/fnmatch.html

The issue is that the modules directory will fail many clang-tidy checks in ways that aren't fixable (since it's not our code). I could remove the filter, but then `scons` won't generate a compilation database usable by `clang-tidy`.

jj-marr replied:
I replaced this with a regex on the `run-clang-tidy` command itself.

soliton- commented on .github/workflows/ci-main.yml:
```diff
@@ -90,10 +90,15 @@ jobs:
+              build cdb
+              # disable all warnings since we have coverage on them in a standard build, and clang-tidy
+              # triggers false positive compiler warnings that clang itself won't
+              run-clang-tidy -use-color -j 2 -extra-arg="-w" '^(?!.*src\/modules|.*build\/).*$'
```
Presumably the backslashes do nothing. Trailing `.*$` does not look useful either.

Anything wrong with `^(?!src/modules/|build/)`?

jj-marr replied:
@soliton- The compilation database uses absolute paths, so the minimum I can reduce it to is `'^(?!.*src/modules/|build/)'`

jj-marr replied:
'^(?!.*src/modules/|.*build/)', actually.

soliton- replied:
How do you get the compilation DB to contain absolute file paths? I didn't manage to.

Oh I see run-clang-tidy puts directory and file together before applying the regex. Makes no sense to me but oh well.

soliton- commented:
Thanks for your contribution!
